File: backend/app/models/finance.py
# today_book/backend/app/models/finance.py
from sqlalchemy import Column, Integer, String, Text, Float, Boolean, Date, DateTime, ForeignKey
from sqlalchemy.orm import relationship
from datetime import datetime, timezone
from app.database import Base
from app.models.types import JSONList


class FinanceAccount(Base):
    __tablename__ = "finance_accounts"

    id = Column(Integer, primary_key=True, autoincrement=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    name = Column(String(100), nullable=False)
    account_type = Column(String(20), nullable=False)  # cash/bank/credit/investment/other
    initial_balance = Column(Float, default=0)
    current_balance = Column(Float, default=0)
    currency = Column(String(10), default="CNY")
    is_active = Column(Boolean, default=True)
    sort_order = Column(Integer, default=0)
    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc))

    user = relationship("User", back_populates="finance_accounts")
    transactions = relationship("FinanceTransaction", foreign_keys="FinanceTransaction.account_id", back_populates="account", cascade="all, delete-orphan")


class FinanceCategory(Base):
    __tablename__ = "finance_categories"

    id = Column(Integer, primary_key=True, autoincrement=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    name = Column(String(100), nullable=False)
    category_type = Column(String(10), nullable=False)  # income/expense
    parent_id = Column(Integer, ForeignKey("finance_categories.id"))
    icon = Column(String(50))
    color = Column(String(20))
    sort_order = Column(Integer, default=0)
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc))

    user = relationship("User", back_populates="finance_categories")
    transactions = relationship("FinanceTransaction", back_populates="category", cascade="all, delete-orphan")
    budgets = relationship("FinanceBudget", back_populates="category", cascade="all, delete-orphan")


class FinanceTransaction(Base):
    __tablename__ = "finance_transactions"

    id = Column(Integer, primary_key=True, autoincrement=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    account_id = Column(Integer, ForeignKey("finance_accounts.id"), nullable=False)
    category_id = Column(Integer, ForeignKey("finance_categories.id"))
    transaction_type = Column(String(10), nullable=False)  # income/expense/transfer
    amount = Column(Float, nullable=False)
    description = Column(Text)
    transaction_date = Column(Date, nullable=False)
    transaction_time = Column(String(8))  # HH:MM:SS
    to_account_id = Column(Integer, ForeignKey("finance_accounts.id"))
    is_recurring = Column(Boolean, default=False)
    tags = Column(JSONList)
    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime, default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))

    user = relationship("User", back_populates="finance_transactions")
    account = relationship("FinanceAccount", foreign_keys=[account_id], back_populates="transactions")
    category = relationship("FinanceCategory", back_populates="transactions")


class FinanceBudget(Base):
    __tablename__ = "finance_budgets"

    id = Column(Integer, primary_key=True, autoincrement=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    category_id = Column(Integer, ForeignKey("finance_categories.id"))
    amount = Column(Float, nullable=False)
    period_type = Column(String(10), nullable=False)  # monthly/yearly
    period_start = Column(Date, nullable=False)
    period_end = Column(Date, nullable=False)
    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime, default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))

    user = relationship("User", back_populates="finance_budgets")
    category = relationship("FinanceCategory", back_populates="budgets")

File: frontend/src/api/finance.ts
/* today_book/frontend/src/api/finance.ts */
import apiClient from './client'

export interface FinanceAccount {
  id: number
  name: string
  account_type: string
  current_balance: number
  initial_balance: number
  currency: string
  is_active: boolean
}

export interface FinanceCategory {
  id: number
  name: string
  category_type: string
  parent_id?: number
  icon?: string
  color?: string
}

export interface FinanceTransaction {
  id: number
  account_id: number
  category_id?: number
  transaction_type: string
  amount: number
  description?: string
  transaction_date: string
  transaction_time?: string
  to_account_id?: number
  tags?: string[] | null
}

export interface FinanceBudget {
  id: number
  category_id?: number
  amount: number
  period_type: string
  period_start: string
  period_end: string
}

export const financeApi = {
  getAccounts: () => apiClient.get<FinanceAccount[]>('/finance/accounts'),
  createAccount: (data: Partial<FinanceAccount>) => apiClient.post('/finance/accounts', data),
  getCategories: (type?: string) => apiClient.get<FinanceCategory[]>('/finance/categories', { params: { cat_type: type } }),
  createCategory: (data: Partial<FinanceCategory>) => apiClient.post('/finance/categories', data),
  getTransactions: (params?: Record<string, unknown>) => apiClient.get<FinanceTransaction[]>('/finance/transactions', { params }),
  createTransaction: (data: Partial<FinanceTransaction>) => apiClient.post('/finance/transactions', data),
  updateTransaction: (id: number, data: Partial<FinanceTransaction>) => apiClient.put(`/finance/transactions/${id}`, data),
  deleteTransaction: (id: number) => apiClient.delete(`/finance/transactions/${id}`),
  getBudgets: () => apiClient.get<FinanceBudget[]>('/finance/budgets'),
  createBudget: (data: Partial<FinanceBudget>) => apiClient.post('/finance/budgets', data),
  getSummary: () => apiClient.get('/finance/reports/summary'),
  getMonthlyReport: (year: number, month: number) => apiClient.get(`/finance/reports/monthly?year=${year}&month=${month}`),
}
